File: Recursion/Lecture 05/CodeStudio/Check Palindrome.cpp
bool isPalindromeNumber(int l, int r, string &s)
{
    // base condition
    // l : left pointer, r : right pointer
    if (l >= r)
        return true;

    // base condition
    // when character at s[left] is not equal to character at s[right] that means it is not a palindrome thus return false
    if (s[l] != s[r])
        return false;

    // recursion call
    return isPalindromeNumber(l + 1, r - 1, s);
}

bool isPalindrome(string &s)
{
    // Write your code here.
    // here we are going to use two pointer approach
    // left pointer starting from first index 0
    // right pointer starting from last index s.size() - 1
    return isPalindromeNumber(0, s.size() - 1, s);
}

/*
    time complexity : O(n)
    space complexity : O(n)
*/

File: Recursion/Lecture 05/CodeStudio/Check Palindrome.py
def isPalindrome(s: str) -> bool:
        
    # helper function to check for palindrome
    def isPalindromeHelper(l, r, s):
        
        # Base case, checking if the left pointer crosses the right pointer, or overlaps then it's a palindrome
        if l >= r:
            return True
        
        # at any moment if the s[left] != s[right] stop and return Fasle
        if s[l] != s[r]:
            return False
        
        return isPalindromeHelper(l+1, r-1, s)
    
    # Recursive call
    return isPalindromeHelper(0, len(s)-1, s)
    
"""
Algorithm
Tc : O(n/2)
Sc : O(1) || O(n/2) stack memory

"""
